include/aspect/melt.h

```diff
@@ -0,0 +1,279 @@
+                                           internal::Assembly::Scratch::StokesSystem<dim>  &scratch,
+                                           internal::Assembly::CopyData::StokesSystem<dim> &data) const;
+
+        /**
+         * Compute the boundary integrals for the Stokes right hand side in
+         * the case of melt migration on a single cell. These boundary terms
+         * are used to describe Neumann boundary conditions for the fluid pressure.
+         */
+        void
+        local_assemble_stokes_system_melt_boundary (const typename DoFHandler<dim>::active_cell_iterator &cell,
+                                                    const unsigned int                                    face_no,
+                                                    const double                                          pressure_scaling,
+                                                    internal::Assembly::Scratch::StokesSystem<dim>       &scratch,
+                                                    internal::Assembly::CopyData::StokesSystem<dim>      &data) const;
+
+        /**
+         * Compute the integrals for the Advection system matrix and right hand side in
+         * the case of melt migration on a single cell.
+         */
+        void
+        local_assemble_advection_system_melt (const typename Simulator<dim>::AdvectionField      &advection_field,
+                                              const double                                        artificial_viscosity,
+                                              internal::Assembly::Scratch::AdvectionSystem<dim>  &scratch,
+                                              internal::Assembly::CopyData::AdvectionSystem<dim> &data) const;
+
+        /**
+         * Compute the residual of the advection system on a single cell in
+         * the case of melt migration.
+         */
+        std::vector<double>
+        compute_advection_system_residual_melt(const typename Simulator<dim>::AdvectionField     &advection_field,
+                                               internal::Assembly::Scratch::AdvectionSystem<dim> &scratch) const;
+
+
+      private:
+        /**
+         * Returns the right hand side of the fluid pressure equation in
+         * the case of melt migration for a single quadrature point.
+         */
+        double
+        compute_fluid_pressure_RHS(const internal::Assembly::Scratch::StokesSystem<dim>  &scratch,
+                                   MaterialModel::MaterialModelInputs<dim> &material_model_inputs,
+                                   MaterialModel::MaterialModelOutputs<dim> &material_model_outputs,
+                                   const unsigned int q_point) const;
+
+        /**
+         * Compute the right-hand side for the advection system index. This is
+         * 0 for the temperature and all of the compositional fields, except for
```

Comment:
Please exchange the "This is 0 ... " by "This function returns 0 for ....". I was wondering why the advection system index should be 0 for temperature and all compositional fields :smile:.